### PR TITLE
Documented a common integration tests problem

### PR DESCRIPTION
**Pull request by yosrym93**
Commit Message:
Documented a common integration tests problem when the response content-length header does not match the actual body length.
Signed-off-by: Yosry Ahmed <yosryahmed@google.com>

Risk Level: N/A
Testing: N/A
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

File: test/integration/README.md
Comment (alyssawilk) [inline test/integration/README.md:103]: I think this could use a tweak - most requests are "sent" without a content length header, but helper functions add them, for example sendRequestAndWaitForResponse will send an end stream bit, indicating if the response is done (and sometimes causing content length to be added)

I think if you remove "(or there is no header at all)" it'd be more accurate.

Comment (alyssawilk) [reply test/integration/README.md:103]: also timeout -> time out

Comment (yosrym93) [reply test/integration/README.md:103]: I faced this problem when I was sending a request using `makeHeaderOnlyRequest` and encoding the response using `upstream_request_->encodeXXX()`. Even when I set `end_stream`, for e.g. by using `upstream_request_->encodeData(_, true)`, the call for `response->waitForEndStream` times out. This happened because I did not include a `content-length` to the encoded headers at all. That's why I added the "no header at all" part.

I guess whether the `content-length` header is added or not depends on the protocol, so it should not be something we depend on. I know this because when I was missing the `content-length` header, the integration test passed for some protocols and timed out for others.

I'd say its better for whoever is writing the test to add a correct `content-length` header, even if some functions/protocols would implicitly add it anyway. That's why I also chose the wording "it could time out" rather than "it will time out".

**Comment by alyssawilk** [comment]
Can you share the code of your previously failing test with me?

I know mismatched content lengths will cause a problem, but if a response
is sent with no content length, Envoy will chunk encode with a terminal
chunk or use the end_stream bit for HTTP/2.  I don't know what went wrong
with your test case but I literally just threw together an integration test
with no response content and it worked fine, so maybe if I understand the
error you encountered I can help you document exactly what the failure mode
is?

**Comment by yosrym93** [comment]
> Can you share the code of your previously failing test with me?

In #12832 that I just submitted recently, the test I added to protocol_integration_test here:
https://github.com/envoyproxy/envoy/pull/12832/files#diff-1d3e9c7609566ad39972e60840c4d384

In the `continue-headers-only-inject-body-filter` code here:
https://github.com/envoyproxy/envoy/pull/12832/files#diff-50d09ef0a52a4f57d86897574f214739

If you comment line 32 (`headers.setContentLength(body_.length())` in `encodeHeaders`), this will cause a time out.

**Comment by yosrym93** [comment]
Interestingly, some protocol combinations pass, some fail, and some timeout. Also in some combinations the `content-length` header is non-existent, while in others a value of 0 is added. Yet there's no direct correlation between this and the time out. For example:
`IPv4_HttpDownstream_HttpUpstream` puts a `content-length` value of 0 but it doesn't time out, it fails with a body mismatch.
`IPv6_Http2Downstream_HttpUpstream` also puts a `content-length` value of 0, but it does time out.

But I noticed that all the cases where no `content-length` header is added at all (null) pass. On the other hand, cases where the the `content-length` is set to 0 (which I guess is a default behavior somewhere) either fail or time out.

**Comment by alyssawilk** [comment]
That's an interesting failure mode for sure.  Envoy totally handles bodies
without the content length header set, so if it's required for
injectEncodedDataToFilterChain I wonder if there's a bug somewhere for that
path?  I don't think I can take a look this week but I'm going to suggest
that 1) it's worth understanding what's going on there for that PR and 2)
it's not a common failure mode (sending no content length in request or
response headers in end to end tests does work, and so I'd prefer we not
document it as something likely to break things as it may cause folks to do
unnecessary work to "fix" it

**Comment by yosrym93** [comment]
I tried it before with `addEncodedData`, same outcome. So I don't think it's about the method the filters uses to add the body. 
1) It seems like something that's happening beyond the HCM and the filter manager, maybe in the codec? I don't know. I have no familiarity whatsoever with code beyond the filter manager and the HCM.
2) I agree. If this is not a common failure mode it should not be documented as a common problem.

Unfortunately I am not able to investigate this. I can open an issue to report the bug, and then someone else might pick it up later. What do you think?

**Comment by alyssawilk** [comment]
Still a bit confused - if your "inject body" test only works with a content length up front, how are you going to combine chunks when you may not know the ultimate size? I think you have to solve that problem before you can land your caching feature.

Either way I'm going to still say that given I've shown that the integration test framework (client and server) work fine without content length, saying you need to have content length seems misleading and I would not like to add that.  I'm happy to call in another maintainer for a second opinion if you feel strongly it should be left in!

**Comment by alyssawilk** [comment]
Ah, per your earlier example

> If you comment line 32 (headers.setContentLength(body_.length()) in encodeHeaders), this will cause a time out.

If you comment out the line it causes problems, but if you replace with
headers.removeContentLength()
it is fine.

My guess is the HCM was setting the content length to 0 in the utility functions, when there was no body
https://github.com/envoyproxy/envoy/blob/master/source/common/http/conn_manager_utility.cc#L93
https://github.com/envoyproxy/envoy/blob/master/source/common/http/conn_manager_utility.cc#L402

So if you didn't *clear* the content length explicitly, you accidentally ended up with the wrong content length.  So again I think the part of your change discussing incorrect content lengths is worth adding, but in this case setting a content length wasn't necessary either.

**Comment by yosrym93** [comment]
> Still a bit confused - if your "inject body" test only works with a content length up front, how are you going to combine chunks when you may not know the ultimate size? I think you have to solve that problem before you can land your caching feature.

The CacheFilter knows the length of the total body up front and adds the content length header accordingly in `encodeHeaders`.

>  So if you didn't clear the content length explicitly, you accidentally ended up with the wrong content length. So again I think the part of your change discussing incorrect content lengths is worth adding, but in this case setting a content length wasn't necessary either.

I will modify the added docs to only mention incorrect content lengths.

**Comment by yosrym93** [comment]
@alyssawilk looking again at `mutateResponseHeaders` utility at:
https://github.com/envoyproxy/envoy/blob/master/source/common/http/conn_manager_utility.cc#L402

It seems like if there's no Content-Length header, the HCM will set it to 0. Isn't this ultimately equivalent to having an incorrect `Content-Length`? (unless the response has no body ofc).

**Comment by alyssawilk** [comment]
Ugh I'm sorry, that was a bad link.
That line is force-setting content length 0 on the upgrade path, not in general
The place we force-set to zero that's relevant to your PR is *this*
https://github.com/envoyproxy/envoy/blob/master/source/common/http/http1/codec_impl.cc#L165

Where we set it to 0 for header-only requests (which the test client was originally sending, before your filter altered things)

**Review by alyssawilk — APPROVED** [review]
LGTM, pending CI churn - thanks for improving our docs!

**Comment by yosrym93** [comment]
> Where we set it to 0 for header-only requests (which the test client was originally sending, before your filter altered things)

Makes sense, thanks!

> The place we force-set to zero that's relevant to your PR is this
https://github.com/envoyproxy/envoy/blob/master/source/common/http/http1/codec_impl.cc#L165

I'm curious why the content-length is set to 0 in that path though. The filter returns `ContinueAndDontEndStream`, which should change `end_stream` to false. I will look into it.

**Comment by yosrym93** [comment]
> The place we force-set to zero that's relevant to your PR is this
https://github.com/envoyproxy/envoy/blob/master/source/common/http/http1/codec_impl.cc#L165

I tried removing the `setContentLength` from the test filter and putting an assert in that path. As I suspected, `end_stream` is false, so it seems like this is not what causes the 0 content-length. Line 165 was never hit.

**Pull request merged**
